File: apps/dashboard-react/src/services/apiClient.ts
import axios, { AxiosInstance } from "axios"

// ------------------------------------------------------------
// Configuration
// ------------------------------------------------------------

const BASE_URL =
  process.env.NEXT_PUBLIC_API_BASE_URL || "http://localhost:8000"

// ------------------------------------------------------------
// Axios instance
// ------------------------------------------------------------

const api: AxiosInstance = axios.create({
  baseURL: BASE_URL,
  timeout: 15000,
  headers: {
    "Content-Type": "application/json",
  },
})

// ------------------------------------------------------------
// Domain types (aligned with backend models)
// ------------------------------------------------------------

export interface EvidenceItem {
  source_model: string
  signal_code: string
  signal_description: string
  confidence: number
  observed_at?: string
}

export interface Recommendation {
  title: string
  rationale: string
  urgency: "LOW" | "MEDIUM" | "HIGH"
  suggested_action?: string
  evidence: EvidenceItem[]
}

export interface VinInterpretation {
  vin: string
  summary: string
  risk_level: "LOW" | "ELEVATED" | "HIGH" | "CRITICAL"
  recommendations: Recommendation[]
  evidence_summary?: Record<string, Record<string, unknown>>
  model_version: string
}

export interface CohortMetric {
  name: string
  value: number
  unit?: string
  description?: string
}

export interface CohortAnomaly {
  title: string
  description: string
  affected_vin_count: number
  severity: string
  related_signals?: string[]
}

export interface CohortInterpretation {
  cohort_id: string
  summary: string
  metrics: CohortMetric[]
  anomalies: CohortAnomaly[]
  risk_distribution?: Record<string, number>
  model_version: string
}

// ------------------------------------------------------------
// API functions
// ------------------------------------------------------------

export async function fetchVinInterpretation(
  vin: string
): Promise<VinInterpretation> {
  const { data } = await api.get<VinInterpretation>(`/vin/${vin}`)
  return data
}

export async function fetchCohortInterpretation(
  cohortId: string
): Promise<CohortInterpretation> {
  const { data } = await api.get<CohortInterpretation>(
    `/cohort/${cohortId}`
  )
  return data
}

export async function createActionPack(payload: {
  subject_type: string
  subject_id: string
  title: string
  executive_summary: string
  recommendations: Recommendation[]
}) {
  const { data } = await api.post("/action-pack", payload)
  return data
}

export async function fetchChatReply(payload: {
  message: string
  context?: Record<string, unknown>
}) {
  const { data } = await api.post<{ reply: string }>("/chat", payload)
  return data.reply
}

export async function exportPdf(payload: {
  subject_type: "vin" | "cohort"
  subject_id: string
}) {
  const { data } = await api.post<ArrayBuffer>("/export/pdf", payload, {
    responseType: "arraybuffer",
  })
  return data
}

export async function recordApproval(payload: {
  subject_type: string
  subject_id: string
  decision: "approve" | "reject" | "escalate"
  comment: string
  decided_by?: string
}) {
  const { data } = await api.post("/approval", payload)
  return data
}

File: apps/dashboard-streamlit/app/services/api_client.py
"""
Streamlit API Client.

Centralized HTTP client for communicating with the
GenAI Predictive Interpreter FastAPI backend.
"""

from __future__ import annotations

import os
import requests
from typing import Dict, Any, List

# ------------------------------------------------------------
# Configuration
# ------------------------------------------------------------

API_BASE_URL = os.getenv(
    "API_BASE_URL",
    "http://localhost:8000",
)

TIMEOUT_SECONDS = 15


# ------------------------------------------------------------
# Low-level request helper
# ------------------------------------------------------------

def _request(
    method: str,
    path: str,
    json: Dict | None = None,
) -> Dict[str, Any]:
    """
    Internal HTTP request helper.
    """
    url = f"{API_BASE_URL}{path}"

    response = requests.request(
        method=method,
        url=url,
        json=json,
        timeout=TIMEOUT_SECONDS,
    )

    if not response.ok:
        raise RuntimeError(
            f"API request failed [{response.status_code}]: {response.text}"
        )

    return response.json()


# ------------------------------------------------------------
# Public API methods
# ------------------------------------------------------------

def fetch_vin_interpretation(vin: str) -> Dict[str, Any]:
    """
    Fetch VIN-level interpretation.
    """
    return _request("GET", f"/vin/{vin}")


def fetch_cohort_interpretation(cohort_id: str) -> Dict[str, Any]:
    """
    Fetch cohort-level interpretation.
    """
    return _request("GET", f"/cohort/{cohort_id}")


def fetch_chat_reply(
    message: str,
    context: Dict[str, Any] | None = None,
) -> str:
    """
    Request a GenAI chat explanation.
    """
    payload = {
        "message": message,
        "context": context,
    }

    result = _request("POST", "/chat", json=payload)
    return result["reply"]


def export_pdf(
    subject_type: str,
    subject_id: str,
) -> bytes:
    """
    Generate and return a PDF report.
    """
    url = f"{API_BASE_URL}/export/pdf"

    response = requests.post(
        url,
        json={
            "subject_type": subject_type,
            "subject_id": subject_id,
        },
        timeout=TIMEOUT_SECONDS,
    )

    if not response.ok:
        raise RuntimeError(
            f"PDF export failed [{response.status_code}]: {response.text}"
        )

    return response.content
